includes/subscriptions/class-wc-payments-subscription-service.php

```diff
@@ -874,6 +878,32 @@ public function maybe_cancel_subscription( $subscription, $new_payment_method )
 		}
 	}
 
+	/**
+	 * Checks if the original subscription mode matches current WooPayments mode.
+	 *
+	 * If the original subscription was payed with WooPayments, but in the mode, that doesn't
+	 * match the current WooPayments mode, we need to throw an exception, to prevent the renewal
+	 * order from being created, as it would fail to be paid.
+	 *
+	 * @param array           $items        The items to be added to the renewal order.
+	 * @param WC_Order        $order        Renewal order.
+	 * @param WC_Subscription $subscription The original subscription.
+	 * @throws Subscription_Mode_Mismatch_Exception
+	 * @return array
+	 */
+	public function check_wcpay_mode_for_subscription( array $items, WC_Order $order, WC_Subscription $subscription ): array {
+		$parent_order = $subscription->get_parent();
+		if ( false !== $parent_order ) {
+			$subscription_mode = $parent_order->get_meta( WC_Payments_Order_Service::WCPAY_MODE_META_KEY );
+			$current_mode      = WC_Payments::mode()->is_test() ? Order_Mode::TEST : Order_Mode::PRODUCTION;
+
+			if ( is_string( $subscription_mode ) && '' !== $subscription_mode && $subscription_mode !== $current_mode ) {
+				throw new Subscription_Mode_Mismatch_Exception( 'The subscription was made in the mode "' . $subscription_mode . '" that does not match the current WooPayments mode "' . $current_mode . '".' );
```

Comment:
Subscriptions displays links to the latest `failed-scheduled-actions` logs, meaning that this message will be easily reachable by shop managers, not just developers. Therefore, I would strongly prefer it localized.

To avoid weird placeholders for translations, I'm fine with having two strings: One for live subscriptions in test mode, and one for test subscriptions in live mode.

Reply:
Thanks for pointing this out! Updated in a30697f9a